signac/contrib/job.py

```diff
@@ -481,14 +491,13 @@ def _init(self, force=False):
             self._check_manifest()
 
     def _check_manifest(self):
-        """Check whether the manifest file is correct (if it exists)."""
+        """Check whether the manifest file exists and is correct."""
         fn_manifest = os.path.join(self._wd, self.FN_MANIFEST)
         try:
             with open(fn_manifest, "rb") as file:
                 assert calc_id(json.loads(file.read().decode())) == self._id
-        except OSError as error:
-            if error.errno != errno.ENOENT:
-                raise error
+        except OSError:
+            raise
```

Comment:
Explanation of my reasoning for this change:
- The `_check_manifest` method is _only_ called in `_init`. Thus it is safe to change its behavior as long as it is consistent with the conditions allowed within the calling context in `_init`.
- At the end of `_init` where the `try` statements succeed and `_check_manifest` is called, the manifest _must_ exist (this makes no sense otherwise, even in cases of filesystem error that are handled in `_init`, because `_check_manifest` is guarded by a `try: ... else:`).
- Thus, it must be safe to remove the exception handling code for cases where the manifest does not exist.

This change of behavior is necessary because I want `_check_manifest` to raise a `FileNotFoundError` (a subclass of `OSError`) in the "early exit" attempt. This "EAFP" approach of raising an error in `_check_manifest`, instead of the "LBYL" approach of checking that the file exists before validating, minimizes the total number of system calls needed for the early exit for optimal performance. ([EAFP vs LBYL](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/))

Reply:
I think you reasoning is sound, but then I would suggest to completely remove lines 499..500 as well, because you are just catching an exception and raising it again so you can just remove it.

Reply:
Good catch. I removed those lines.